presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java

```diff
@@ -492,6 +493,99 @@ public PlanNode visitJoin(JoinNode node, RewriteContext<Expression> context)
             return output;
         }
 
+        @Override
+        public PlanNode visitSpatialJoin(SpatialJoinNode node, RewriteContext<Expression> context)
+        {
+            Expression inheritedPredicate = context.get();
```

Comment:
@kokosing `visitSpatialJoin` is a copy + delete of the `visitJoin`; I looked to see if I could reduce copy-paste, but I can't see how. Do you have any specific suggestions?

Reply:
I thought about unifying `InnerJoinPushDownResult` and `OuterJoinPushDownResult`. They look very similar, that way we could reuse the code around the `switch`, but I am not sure if it is good refactor.

Let's leave it as it is.

Reply:
Is there any good reason why don't we want to have an hirerachy of joins?

Like: 

```
                     [JoinNode]
                     - type
                     - left
                     - rigth
                     - outputSymbols
           |             |
[EquiJoinNode]           [SpatialJoinNode]
- criteria                      - filter
- leftHashSymbol
- rightHashSymbol
- distributionType
- optional filter
```

(and maybe similar for `IndexJoinNode`)

This hirerarchy allows to remove most of the copy paste of the methods that do not return a new join node. Also it may help to remove some copy-paste for the rest of the methods.

CC: @kokosing @martint

Reply:
I am ok to this. Let's save this refactor for another PR.